File: GakumasuCalc/Models/StatusValues.cs
namespace GakumasuCalc.Models;

public class StatusValues
{
    public int Vo { get; set; }
    public int Da { get; set; }
    public int Vi { get; set; }

    public int Total => Vo + Da + Vi;

    public StatusValues() { }

    public StatusValues(int vo, int da, int vi)
    {
        Vo = vo;
        Da = da;
        Vi = vi;
    }

    public static StatusValues Zero => new(0, 0, 0);

    public StatusValues Add(StatusValues other)
    {
        return new StatusValues(Vo + other.Vo, Da + other.Da, Vi + other.Vi);
    }

    public StatusValues MultiplyAndFloor(double factor)
    {
        return new StatusValues(
            (int)Math.Floor(Vo * factor),
            (int)Math.Floor(Da * factor),
            (int)Math.Floor(Vi * factor));
    }

    public StatusValues Clone() => new(Vo, Da, Vi);

    public override string ToString() => $"Vo:{Vo} Da:{Da} Vi:{Vi} (Total:{Total})";
}

File: web/src/utils/statusValues.ts
import type { StatusValues } from '../types/models';

export function sv(vo: number, da: number, vi: number): StatusValues {
  return { vo, da, vi };
}

export const ZERO: StatusValues = { vo: 0, da: 0, vi: 0 };

export function svZero(): StatusValues {
  return { vo: 0, da: 0, vi: 0 };
}

export function svAdd(a: StatusValues, b: StatusValues): StatusValues {
  return { vo: a.vo + b.vo, da: a.da + b.da, vi: a.vi + b.vi };
}

export function svMultiplyAndFloor(s: StatusValues, factor: number): StatusValues {
  return {
    vo: Math.floor(s.vo * factor),
    da: Math.floor(s.da * factor),
    vi: Math.floor(s.vi * factor),
  };
}

export function svClone(s: StatusValues): StatusValues {
  return { vo: s.vo, da: s.da, vi: s.vi };
}

export function svTotal(s: StatusValues): number {
  return s.vo + s.da + s.vi;
}
